solr/test-framework/src/java/org/apache/solr/util/EmbeddedSolrServerTestRule.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.util;
+
+import static org.apache.solr.SolrTestCaseJ4.DEFAULT_TEST_COLLECTION_NAME;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.lucene.tests.util.LuceneTestCase;
+import org.apache.solr.client.solrj.embedded.EmbeddedSolrServer;
+import org.apache.solr.client.solrj.embedded.EmbeddedSolrServer.RequestWriterSupplier;
+import org.apache.solr.client.solrj.impl.HttpClientUtil;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.core.CoreDescriptor;
+import org.apache.solr.core.NodeConfig;
+import org.apache.solr.update.UpdateShardHandlerConfig;
+
+/** TODO NOCOMMIT document */
+public class EmbeddedSolrServerTestRule extends SolrClientTestRule {
+
+  private EmbeddedSolrServer client = null;
+
+  public Builder build() {
+    return new Builder();
+  }
+
+  public class Builder {
+    private Path solrHome; // mandatory
+    private Path dataDir;
+    private String collectionName = DEFAULT_TEST_COLLECTION_NAME;
+    private String configFile;
+    private String schemaFile;
+    private RequestWriterSupplier requestWriterSupplier = RequestWriterSupplier.JavaBin;
+
+    public Builder setSolrHome(Path solrHome) {
+      this.solrHome = solrHome;
+      return this;
+    }
+
+    public Builder useTempDataDir() {
+      this.dataDir = LuceneTestCase.createTempDir("data-dir");
+      return this;
+    }
+
+    public Builder setSchemaFile(String schemaFile) {
```

Comment:
[0] Eric Pugh has been working recently on a Builder pattern for creating Solr clients that has methods with names like: `withFoo`, `withBar`, etc.

I love that you're consistent across this Builder here with using `set`.  But is there any value in being consistent with the SolrClient builders and using `with` instead of `set`?  (Maybe there's not - solrj and solr-test-framework are different animals.  Just bringing it up for conversation in case it hasn't been raised yet.)

Reply:
Should use "with" in general -- basically whatever reads well.